README.md
# misaem

### Introduction

`misaem` is an implementation of methodology which performs statistical inference for logistic regression model with missing data. This method is based on likelihood, including:

1. Estimate the parameters of logistic regression by a stochastic approximation version of EM algorithm;
2. Estimation of parameters' variance based one Louis formula;
3. Model selection procedure based on BIC;
4. Prediction on a test set which may contain missing values.

### Installation of package 
Now you can install the package **misaem** from CRAN. 
```{r}
install.packages("misaem")
 ```
### Using the misaem package
Basically,

1. `miss.saem` contains the procedure of estimation for parameters, as well as their variance, and observed likelihood.
2. `model_selection` aims at selecting a best model according to BIC.
3. `pred_saem` performs prediction on a test set which may contain missing values.

For more details, You can find the vignette, which illustrate the basic and further usage of misaem package:
```{r}
library(misaem)
vignette('misaem')
 ```

### Reference
Stochastic Approximation EM for Logistic regression with missing values (2018, Jiang W., Josse J., Lavielle M., Traumabase group)" [arxiv link](https://arxiv.org/abs/1805.04602).




NEWS.md
# misaem 0.9.2

A minor release mainly fixing bug:

* Replace `if (class(obj) == "data.frame")  { ..... }` with `if (is(obj, "data.frame"))  { ..... }`. As within the formal S4 class system, a matrix `obj` could well be of a class that extends matrix, such as "matrix" "array" (length = 2).

# misaem 0.9.1

A minor release mainly fixing bugs and typos:

* Fix a bug in `model_selection`, now it can correctly perform model selection if the full model is the best model.

* In `pred_saem`, two methods for prediction of test set with missingness are provided.

* Fix some typos in the vignettes. The length of codes now fits the page wide of html.

* Delete unused Imports ‘magrittr’ in DESCRIPTION file.

* Change the index of vignitte from 'SAEM' to 'misaem tutorial'.

* Update README.md.


R/pred_saem.R
#' pred_saem
#'
#' Prediction on test with missing values for the logistic regression model.
#' @param X.test Design matrix in test set.
#' @param beta.saem Estimated \eqn{\beta}{\beta} by SAEM.
#' @param sig2.saem Estimated \eqn{\Sigma}{\Sigma} by SAEM.
#' @param mu.saem Estimated \eqn{\mu}{\mu} by SAEM.
#' @param seed  An integer as a seed set for the radom generator. The default value is 200.
#' @param method The name of method to deal with missing values in test set. It can be 'map'(maximum a posteriori) or 'impute' (imputation by conditional expectation). Default is 'map'.
#' @return
#' \item{pr.saem}{The prediction result for logistic regression: the probability of response y=1.}
#' @import mvtnorm stats MASS
#' @importFrom methods is
#' @examples
#' # Generate dataset
#' N <- 100  # number of subjects
#' p <- 3     # number of explanatory variables
#' mu.star <- rep(0,p)  # mean of the explanatory variables
#' Sigma.star <- diag(rep(1,p)) # covariance
#' beta.star <- c(1, 1,  0) # coefficients
#' beta0.star <- 0 # intercept
#' beta.true = c(beta0.star,beta.star)
#' X.complete <- matrix(rnorm(N*p), nrow=N)%*%chol(Sigma.star) +
#'               matrix(rep(mu.star,N), nrow=N, byrow = TRUE)
#' p1 <- 1/(1+exp(-X.complete%*%beta.star-beta0.star))
#' y <- as.numeric(runif(N)<p1)

#' # Generate missingness
#' p.miss <- 0.10
#' patterns <- runif(N*p)<p.miss #missing completely at random
#' X.obs <- X.complete
#' X.obs[patterns] <- NA
#'
#' # SAEM
#' list.saem = miss.saem(X.obs,y)
#'
#' # Generate test set with missingness
#' Nt = 50
#' X.test <- matrix(rnorm(Nt*p), nrow=Nt)%*%chol(Sigma.star)+
#'               matrix(rep(mu.star,Nt), nrow=Nt, byrow = TRUE)
#' p1 <- 1/(1+exp(-X.test%*%beta.star-beta0.star))
#' y.test <- as.numeric(runif(Nt)<p1)
#'
#' # Prediction on test set
#' pr.saem <- pred_saem(X.test, list.saem$beta, list.saem$mu, list.saem$sig2)
#' pred.saem <- (pr.saem>0.5)*1
#' table(y.test, pred.saem)
#' @export


pred_saem = function(X.test,beta.saem,mu.saem,sig2.saem,seed=200,method='map'){

  #judge
  if (is(X.test, "data.frame")){
    X.test <- as.matrix(X.test)
  }
  if (sum(sapply(X.test, is.numeric)) < ncol(X.test)) {
    stop("Error: the variables should be numeric.")
  }
  if (method == "MAP" | method == "Map") {
    method <- "map"
  }
  if (method == "Impute" | method == "IMPUTE") {
    method <- "impute"
  }
  set.seed(seed)
  rindic = as.matrix(is.na(X.test))

  if(method=='impute'){
    for(i in 1:dim(X.test)[1]){
      if(sum(rindic[i,])!=0){
        miss_col = which(rindic[i,]==TRUE)
        x2 = X.test[i,-miss_col]
        mu1 = mu.saem[miss_col]
        mu2 = mu.saem[-miss_col]
        sigma11 = sig2.saem[miss_col,miss_col]
        sigma12 = sig2.saem[miss_col,-miss_col]
        sigma22 = sig2.saem[-miss_col,-miss_col]
        sigma21 = sig2.saem[-miss_col,miss_col]
        mu_cond = mu1+sigma12 %*% solve(sigma22) %*% (x2-mu2)
        X.test[i,miss_col] =mu_cond
      }
    }
    tmp <- as.matrix(cbind.data.frame(rep(1,dim(X.test)[1]),X.test)) %*% as.matrix(beta.saem)
    pr.saem <- 1/(1+(1/exp(tmp)))
  }else if(method=='map'){
    pr2 =rep(0,dim(X.test)[1])
    mc.size = 100
    X.test = data.matrix(X.test)
    for(i in 1:dim(X.test)[1]){
      x=X.test[i,]
      if(sum(rindic[i,])==0){
        pr2[i]=log_reg(y=1,x=c(1,x),beta.saem,iflog=FALSE)
      }
      else{
        miss_col = which(rindic[i,]==TRUE)
        x2 = X.test[i,-miss_col]
        mu1 = mu.saem[miss_col]
        mu2 = mu.saem[-miss_col]
        sigma11 = sig2.saem[miss_col,miss_col]
        sigma12 = sig2.saem[miss_col,-miss_col]
        sigma22 = sig2.saem[-miss_col,-miss_col]
        sigma21 = sig2.saem[-miss_col,miss_col]
        mu_cond = mu1+sigma12 %*% solve(sigma22)%*%(x2-mu2)
        sigma_cond = sigma11 - sigma12 %*% solve(sigma22) %*% sigma21
        x1_all=mvrnorm(n = mc.size, mu_cond, sigma_cond, tol = 1e-6, empirical = FALSE, EISPACK = FALSE)
        p1=0
        for(j in 1:mc.size){
          x[miss_col] =x1= x1_all[j,]
          p1 = p1 + log_reg(y=1,x=c(1,x),beta.saem,iflog=FALSE)
        }
        pr2[i] =p1/mc.size
      }
    }
    pr.saem = as.matrix(pr2)
  } else {
    stop("Error: There is no such method. Method should be 'map' or 'impute'. ")
  }
  return(pr.saem)
}





R/likelihood_saem.R
#' likelihood_saem
#'
#' Used in main function miss.saem. Calculate the observed log-likelihood for logistic regression model with missing data, using Monte Carlo version of Louis formula.
#' @param beta Estimated parameter of logistic regression model.
#' @param mu Estimated parameter \eqn{\mu}.
#' @param Sigma Estimated parameter \eqn{\Sigma}.
#' @param Y Response vector \eqn{N \times 1}{N * 1}
#' @param X.obs Design matrix with missingness \eqn{N \times p}{N * p}
#' @param rindic Missing pattern of X.obs. If a component in X.obs is missing, the corresponding position in rindic is 1; else 0.
#' @param whichcolXmissing The column index in covariate containing at least one missing observation.
#' @param mc.size Monte Carlo sampling size.
#' @import mvtnorm stats MASS
#' @return Observed log-likelihood.
#' @examples
#' # Generate dataset
#' N <- 50  # number of subjects
#' p <- 3     # number of explanatory variables
#' mu.star <- rep(0,p)  # mean of the explanatory variables
#' Sigma.star <- diag(rep(1,p)) # covariance
#' beta.star <- c(1, 1,  0) # coefficients
#' beta0.star <- 0 # intercept
#' beta.true = c(beta0.star,beta.star)
#' X.complete <- matrix(rnorm(N*p), nrow=N)%*%chol(Sigma.star) +
#'               matrix(rep(mu.star,N), nrow=N, byrow = TRUE)
#' p1 <- 1/(1+exp(-X.complete%*%beta.star-beta0.star))
#' y <- as.numeric(runif(N)<p1)

#' # Generate missingness
#' p.miss <- 0.10
#' patterns <- runif(N*p)<p.miss #missing completely at random
#' X.obs <- X.complete
#' X.obs[patterns] <- NA
#'
#' # Observed log-likelihood
#' ll_obs = likelihood_saem(beta.true,mu.star,Sigma.star,y,X.obs)
#' @export


likelihood_saem = function(beta,mu,Sigma,Y,X.obs,rindic=as.matrix(is.na(X.obs)),whichcolXmissing=(1:ncol(rindic))[apply(rindic,2,sum)>0],mc.size=2){
  n=dim(X.obs)[1]
  p=dim(X.obs)[2]
  lh=0
  for(i in 1:n){
    y=Y[i]
    x=X.obs[i,]
    if(sum(rindic[i,])==0){
      lr = log_reg(y=y,x=c(1,x),beta,iflog=TRUE)
      nm = dmvnorm(x, mean = mu, sigma = Sigma, log = TRUE)
      lh=lh+lr+nm
    }
    else{
      miss_col = which(rindic[i,]==TRUE)
      x2 = x[-miss_col]
      mu1 = mu[miss_col]
      mu2 = mu[-miss_col]
      sigma11 = Sigma[miss_col,miss_col]
      sigma12 = Sigma[miss_col,-miss_col]
      sigma22 = Sigma[-miss_col,-miss_col]
      sigma21 = Sigma[-miss_col,miss_col]
      mu_cond = mu1+sigma12 %*% solve(sigma22)%*%(x2-mu2)
      sigma_cond = sigma11 - sigma12 %*% solve(sigma22) %*% sigma21
      x1_all=mvrnorm(n = mc.size, mu_cond, sigma_cond, tol = 1e-6, empirical = FALSE, EISPACK = FALSE)
      lh_mis1=0
      for(j in 1:mc.size){
        x[miss_col] =x1= x1_all[j,]
        lh_mis1 = lh_mis1 + log_reg(y=y,x=c(1,x),beta,iflog=FALSE)
      }
      lr = log(lh_mis1/mc.size)
      if(length(x2)>1){nm = dmvnorm(x2, mean = mu2, sigma = sigma22, log = TRUE)}
      if(length(x2)==1){nm = dnorm(x2, mean=mu2, sd=sqrt(sigma22), log = TRUE)}
      lh = lh + lr + nm
    }
  }
  return(ll=lh)
}


R/model_selection.R
#' model_selection
#'
#' Model selection for the logistic regression model with missing data.
#' @param X.obs Design matrix with missingness \eqn{N \times p}{N * p}
#' @param y Response vector \eqn{N \times 1}{N * 1}
#' @param seed  An integer as a seed set for the radom generator. The default value is 200.
#' @return A list with components
#' \item{subset_choose}{The index of variables included in the best model selected.}
#' \item{beta}{Estimated \eqn{\beta}{\beta} for the best model.}
#' \item{sig2}{Estimated \eqn{\Sigma}{\Sigma} for the best model.}
#' \item{mu}{Estimated \eqn{\mu}{\mu} for the best model.}
#' @import mvtnorm stats
#' @importFrom methods is
#' @examples
#' # Generate dataset
#' N <- 20  # number of subjects
#' p <- 3     # number of explanatory variables
#' mu.star <- rep(0,p)  # mean of the explanatory variables
#' Sigma.star <- diag(rep(1,p)) # covariance
#' beta.star <- c(1, 1,  0) # coefficients
#' beta0.star <- 0 # intercept
#' beta.true = c(beta0.star,beta.star)
#' X.complete <- matrix(rnorm(N*p), nrow=N)%*%chol(Sigma.star) +
#'               matrix(rep(mu.star,N), nrow=N, byrow = TRUE)
#' p1 <- 1/(1+exp(-X.complete%*%beta.star-beta0.star))
#' y <- as.numeric(runif(N)<p1)

#' # Generate missingness
#' p.miss <- 0.10
#' patterns <- runif(N*p)<p.miss #missing completely at random
#' X.obs <- X.complete
#' X.obs[patterns] <- NA

#' # model selection for SAEM
#' list.saem.select = model_selection(X.obs,y)
#' print(list.saem.select$subset_choose)
#' print(list.saem.select$beta)
#' @export


model_selection = function(X.obs,y,seed=200){
  #judge

  if (is(X.obs, "data.frame")){
    X.obs <- as.matrix(X.obs)
  }
  if (sum(sapply(X.obs, is.numeric)) < ncol(X.obs)) {
    stop("Error: the variables should be numeric.")
  }
  if (sum(y==1) +  sum(y==0) < nrow(X.obs)) {
    stop("Error: y must be coded by 0 or 1, and there is no missing data in y.")
  }

  set.seed(seed)
  N=dim(X.obs)[1]
  p=dim(X.obs)[2]
  subsets=combinations(p)
  ll = matrix(-Inf,nrow=p,ncol=p)
  subsets1 = subsets[rowSums(subsets)==1,]
  for (j in 1:(nrow(subsets1))){
    pos_var=which(subsets1[j,]==1)
    list.saem.subset=miss.saem(data.matrix(X.obs),y,pos_var,maxruns=500,tol_em=1e-7,nmcmc=2,print_iter=FALSE,ll_obs_cal=TRUE)
    ll[1,j] = list.saem.subset$ll
  }
  id = BIC = rep(0,p)
  subsetsi=subsets1
  SUBSETS = matrix(-Inf,nrow=p,ncol=p)
  for(i in 2:p){
    nb.x = i-1
    nb.para = (nb.x + 1) + p + p*p
    id[i-1] = d = which.max(ll[i-1,])
    pos_var=which(subsetsi[d,]==1)
    BIC[i-1] = -2*ll[i-1,d]+ nb.para * log(N)
    SUBSETS[i-1,]=subsetsi[d,]
    if(i==2){subsetsi = subsets[(rowSums(subsets)==i) & (subsets[,pos_var]==i-1),]}
    if(i>2){subsetsi = subsets[(rowSums(subsets)==i) & (rowSums(subsets[,pos_var])==i-1),]}
    if(i<p){
      for (j in 1:(nrow(subsetsi))){
        pos_var=which(subsetsi[j,]==1)
        list.saem.subset=miss.saem(data.matrix(X.obs),y,pos_var,maxruns=500,tol_em=1e-7,nmcmc=2,print_iter=FALSE,ll_obs_cal=TRUE)
        ll[i,j] = list.saem.subset$ll
      }
    }
  }
  SUBSETS[p,]=rep(1,p)
  list.saem.subset=miss.saem(data.matrix(X.obs),y,1:p,maxruns=500,tol_em=1e-7,nmcmc=2,print_iter=FALSE,ll_obs_cal=TRUE)
  ll[p,1] = list.saem.subset$ll
  nb.x = p
  nb.para = (nb.x + 1) + p + p*p
  BIC[p] = -2*ll[p,1]+ nb.para * log(N)
  subset_choose = which(SUBSETS[which.min(BIC),]==1)
  list.saem.subset=miss.saem(data.matrix(X.obs),y,subset_choose,maxruns=500,tol_em=1e-7,k1=2,nmcmc=2,print_iter=TRUE)
  beta.saem.train = list.saem.subset$beta
  mu.saem.train = list.saem.subset$mu
  sig2.saem.train = list.saem.subset$sig2
  return(list(subset_choose=subset_choose,beta=beta.saem.train,sig2=sig2.saem.train,mu=mu.saem.train))
}


R/miss.saem.R
#' miss.saem
#'
#' This function uses algorithm SAEM to fit the logistic regression model with missing data.
#' @param X.obs Design matrix with missingness \eqn{N \times p}{N * p}
#' @param y Response vector \eqn{N \times 1}{N * 1}
#' @param pos_var Index of selected covariates. The default is pos_var = 1:ncol(X.obs).
#' @param maxruns Maximum number of iterations. The default is maxruns = 500.
#' @param tol_em The tolerance to stop SAEM. The default is tol_em = 1e-7.
#' @param nmcmc The MCMC length. The default is nmcmc = 2.
#' @param tau Rate \eqn{\tau}{\tau} in the step size \eqn{(k-k_{1})^{-\tau}}{(k-k1)^(-\tau)}. The default is tau = 1.
#' @param k1 Number of first iterations \eqn{k_{1}}{k1} in the step size \eqn{(k-k_{1})^{-\tau}}{(k-k1)^(-\tau)}. The default is k1=50.
#' @param seed An integer as a seed set for the random generator. The default value is 200.
#' @param print_iter If TRUE, miss.saem will print the estimated parameters in each iteration of SAEM.
#' @param var_cal If TRUE, miss.saem will calculate the variance of estimated parameters.
#' @param ll_obs_cal If TRUE, miss.saem will calculate the observed log-likelihood.
#' @return A list with components
#' \item{mu}{Estimated \eqn{\mu}{\mu}.}
#' \item{sig2}{Estimated \eqn{\Sigma}{\Sigma}.}
#' \item{beta}{Estimated \eqn{\beta}{\beta}.}
#' \item{time_run}{Execution time.}
#' \item{seqbeta}{Sequence of \eqn{\beta}{\beta} estimated in each iteration.}
#' \item{seqbeta_avg}{Sequence of \eqn{\beta}{\beta} with averaging in each iteration.}
#' \item{ll}{Observed log-likelihood.}
#' \item{var_obs}{Estimated variance for estimated parameters.}
#' \item{std_obs}{Estimated standard error for estimated parameters.}
#' @import mvtnorm stats
#' @importFrom methods is
#' @examples
#' # Generate dataset
#' N <- 100  # number of subjects
#' p <- 3     # number of explanatory variables
#' mu.star <- rep(0,p)  # mean of the explanatory variables
#' Sigma.star <- diag(rep(1,p)) # covariance
#' beta.star <- c(1, 1,  0) # coefficients
#' beta0.star <- 0 # intercept
#' beta.true = c(beta0.star,beta.star)
#' X.complete <- matrix(rnorm(N*p), nrow=N)%*%chol(Sigma.star) +
#'               matrix(rep(mu.star,N), nrow=N, byrow = TRUE)
#' p1 <- 1/(1+exp(-X.complete%*%beta.star-beta0.star))
#' y <- as.numeric(runif(N)<p1)

#' # Generate missingness
#' p.miss <- 0.10
#' patterns <- runif(N*p)<p.miss #missing completely at random
#' X.obs <- X.complete
#' X.obs[patterns] <- NA
#'
#' # SAEM
#' list.saem = miss.saem(X.obs,y)
#' print(list.saem$beta)
#' @export

miss.saem <- function(X.obs,y,pos_var=1:ncol(X.obs),maxruns=500,tol_em=1e-7,nmcmc=2,tau=1,k1=50, seed=200, print_iter=TRUE, var_cal=FALSE, ll_obs_cal=FALSE) {
  set.seed(seed)

  #judge
  if (is(X.obs, "data.frame")){
    X.obs <- as.matrix(X.obs)
  }
  if (sum(sapply(X.obs, is.numeric)) < ncol(X.obs)) {
    stop("Error: the variables should be numeric.")
  }
  if (sum(y==1) +  sum(y==0) < nrow(X.obs)) {
    stop("Error: y must be coded by 0 or 1, and there is no missing data in y.")
  }

  if (sum(pos_var %in% 1:ncol(X.obs)) < length(pos_var))  {
    stop("Error: index of selected variables must be in the range of covariates.")
  }

  if (length(unique(pos_var)) != length(pos_var)){
    stop("Error: index of selected variables must not be repeated.")
  }

  p=ncol(X.obs)

  #delete rows completely missing
  if(any(apply(is.na(X.obs),1,sum)==p)){
    i_allNA=which(apply(is.na(X.obs),1,sum)==p)
    X.obs = X.obs[-i_allNA,]
    y = y[-i_allNA]
  }
  if(any((is.na(y))==TRUE)){
    i_YNA=which(is.na(y)==TRUE)
    X.obs = X.obs[-i_YNA,]
    y = y[-i_YNA]
  }
  n=length(y)


  rindic = as.matrix(is.na(X.obs))
  if(sum(rindic)>0){
    whichcolmissing = (1:ncol(rindic))[apply(rindic,2,sum)>0]
    missingcols = length(whichcolmissing)
  }
  if(sum(rindic)==0){missingcols=0}


  ptm <- Sys.time()
  if(missingcols>0){
    k=0
    cstop=0.1
    seqbeta = matrix(NA,nrow=ncol(X.obs)+1,ncol=(maxruns+1))
    seqbeta_avg = matrix(NA,nrow=ncol(X.obs)+1,ncol=(maxruns+1))

    X.mean = X.obs
    for(i in 1:ncol(X.mean)){
      X.mean[is.na(X.mean[,i]), i] <- mean(X.mean[,i], na.rm = TRUE)
    }
    X.sim <- X.mean

    mu = apply(X.mean,2,mean)
    Sigma = var(X.mean)*(n-1)/n
    beta= rep(0,p+1)
    beta[c(1,pos_var+1)]= glm(y~ X.mean[,pos_var],family=binomial(link='logit'))$coef

    while ((cstop>tol_em)*(k<maxruns)|(k<20)){
      k = k+1
      beta.old = beta

      if(k <k1){gamma <- 1}else{gamma <- 1/(k-(k1-1))^tau}

      S.inv <- solve(Sigma)

      for (i in (1:n)) {
        jna <- which(is.na(X.obs[i,]))
        njna <- length(jna)
        if (njna>0) {
          xi <- X.sim[i,]
          Oi <- solve(S.inv[jna,jna])
          mi <- mu[jna]
          lobs <- beta[1]
          if (njna<p) {
            jobs <- setdiff(1:p,jna)
            mi <- mi - (xi[jobs] - mu[jobs])%*%S.inv[jobs,jna]%*%Oi
            lobs <- lobs + sum(xi[jobs]*beta[jobs+1])
          }

          cobs <- exp(lobs)
          if(cobs==0){cobs=.Machine$double.xmin}
          if(cobs==Inf){cobs=.Machine$double.xmax}

          xina <- xi[jna]
          betana <- beta[jna+1]
          for (m in (1:nmcmc)) {
            xina.c <- mi + rnorm(njna)%*%chol(Oi)

            if (y[i]==1)
              alpha <- (1+exp(-sum(xina*betana))/cobs)/(1+exp(-sum(xina.c*betana))/cobs)
            else
              alpha <- (1+exp(sum(xina*betana))*cobs)/(1+exp(sum(xina.c*betana))*cobs)
            if (runif(1) < alpha){
              xina <- xina.c
            }
          }
          X.sim[i,jna] <- xina
        }
      }
      beta_new= rep(0,p+1)
      beta_new[c(1,pos_var+1)]= glm(y~ X.sim[,pos_var],family=binomial(link='logit'))$coef

      beta <- (1-gamma)*beta + gamma*beta_new
      cstop = sum((beta-beta.old)^2)

      mu <- (1-gamma)*mu + gamma*colMeans(X.sim)
      Sigma <- (1-gamma)*Sigma + gamma*cov(X.sim)

      seqbeta[,k]=beta.old

      if(k==1){
        seqbeta_avg[,k]=beta.old
      }else{
        seqbeta_avg[,k]= 1/k*rowSums(seqbeta[,1:k])
      }

      if(print_iter==TRUE & k %% 10 == 0){
        cat(sprintf('iteration = %i ', k))
        cat(sprintf('beta ='),beta,'\n')
        cat(sprintf('Distance from last iteration ='),cstop,'\n')
      }
    }
    var_obs = ll = std_obs =NULL
    if(var_cal==TRUE){
      var_obs = louis_lr_saem(beta,mu,Sigma,y,X.obs,pos_var,rindic,whichcolmissing,mc.size=1000)
      std_obs <- sqrt(diag(var_obs))
    }
    if(ll_obs_cal==TRUE){
      ll= likelihood_saem(beta,mu,Sigma,y,X.obs,rindic,whichcolmissing,mc.size=1000)
    }
  }
  if(missingcols==0){
    X.obs = matrix(X.obs,nrow=n)
    data.complete <- data.frame(y=y,X.obs)
    model.complete <- glm(y ~.,family=binomial(link='logit'),data=data.complete)
    mu = apply(X.obs,2,mean)
    Sigma = var(X.obs)*(n-1)/n
    beta <- model.complete$coefficients
    var_obs = ll = ll1 =ll2= std_obs =seqbeta_avg= seqbeta=NULL
    if(var_cal==TRUE){
      P <- predict(model.complete, type = "response")
      W <- diag(P*(1-P))
      X <- model.matrix(model.complete)

      var_obs <- solve(t(X)%*%W%*%X)
      std_obs <- sqrt(diag(var_obs))
    }
    if(ll_obs_cal==TRUE){
      ll = likelihood_saem(beta,mu,Sigma,y,X.obs,rindic,whichcolmissing,mc.size=1000)
    }
  }
  time_run=Sys.time() - ptm
  return(list(mu=mu, sig2=Sigma, beta=beta,time_run=time_run,seqbeta=seqbeta,seqbeta_avg=seqbeta_avg,ll=ll,var_obs=var_obs,std_obs=std_obs))
}
